### PR TITLE
issue #8: hibernate issues fixes

### PR DESCRIPTION
**Pull request by tonikromik**

### CODE REVIEW

**Review by financeasist — CHANGES_REQUESTED** [review]

File: src/main/java/com/petstore/controller/StoreController.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/store")
+@Validated
+public class StoreController {
+
+    private final StoreService ordersService;
+    private final OrdersMapper ordersMapper;
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/controller/StoreController.java:20]: moved

File: src/main/java/com/petstore/controller/UserController.java [hunk 2]
```diff
@@ -13,37 +13,38 @@
...
 @RestController
+@RequestMapping("/user")
 @Validated
 @RequiredArgsConstructor
 public class UserController {
 
     private final UserService userService;
 
-    @PostMapping("/user")
+    @PostMapping()
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/controller/UserController.java:23]: removed

File: src/main/java/com/petstore/controller/StoreController.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/store")
+@Validated
+public class StoreController {
+
+    private final StoreService ordersService;
```
Comment (tonikromik) [inline src/main/java/com/petstore/controller/StoreController.java:19]: renamed

File: src/main/java/com/petstore/dto/CategoryDTO.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import lombok.*;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+//@Builder
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/dto/CategoryDTO.java:11]: refactored

File: src/main/java/com/petstore/dto/UserDTO.java [hunk 5]
```diff
@@ -9,7 +9,7 @@
 @Builder
 @NoArgsConstructor
 @AllArgsConstructor
-public class UserDTO {
+public class UserDTO extends BaseDTO{
 
     private Long id;
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/dto/UserDTO.java:14]: fixed

File: src/main/java/com/petstore/entity/Category.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "categories_seq")
+    @SequenceGenerator(name = "categories_seq", sequenceName = "categories_seq", initialValue = 4)
+    private Long id;
+
+    @Column(name = "category_name", nullable = false)
+    private String name;
+
+    @OneToMany(mappedBy = "category", cascade = {CascadeType.PERSIST, CascadeType.MERGE})
+    private List<Pet> pets = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/entity/Category.java:28]: done

File: src/main/java/com/petstore/entity/Orders.java [hunk 7]
```diff
@@ -11,35 +11,34 @@
 @NoArgsConstructor
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/entity/Orders.java:11]: refactored

File: src/main/java/com/petstore/entity/Pet.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+public class Pet extends BaseEntity {
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "pets_seq")
+    @SequenceGenerator(name = "pets_seq", sequenceName = "pets_seq", initialValue = 8)
+    private Long id;
+
+    @Column(name = "pet_name", nullable = false)
+    private String name;
+
+    @ManyToOne(fetch = FetchType.LAZY, cascade = {CascadeType.PERSIST, CascadeType.MERGE})
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/entity/Pet.java:26]: done

File: src/main/java/com/petstore/entity/Tag.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.List;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@Entity
+@Table(name = "tags")
+public class Tag extends BaseEntity {
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "tags_seq")
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/entity/Tag.java:19]: done

File: src/main/java/com/petstore/entity/User.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    @Column(name = "user_name", unique = true, nullable = false)
+    private String userName;
+
+    @Column(name = "first_name")
+    private String firstName;
+
+    @Column(name = "last_name")
+    private String lastName;
+
+    @Column(name = "email", nullable = false)
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/entity/User.java:28]: done

File: src/main/java/com/petstore/mapper/CategoryMapper.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+import com.petstore.dto.CategoryDTO;
+import com.petstore.entity.Category;
+import org.mapstruct.Mapper;
+
+@Mapper(componentModel = "spring")
+public interface CategoryMapper extends GenericMapper<CategoryDTO, Category> {
+//    @Override
+////    @Mapping(target = "pets", ignore = true)
+//    void updateProperties(CategoryDTO dto, @MappingTarget Category entity);
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/mapper/CategoryMapper.java:11]: cleaned and refactored

File: src/main/java/com/petstore/mapper/TagMapper.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package com.petstore.mapper;
+
+import com.petstore.dto.TagDTO;
+import com.petstore.entity.Tag;
+import org.mapstruct.Mapper;
+
+@Mapper(componentModel = "spring", uses = {PetMapper.class})
+public interface TagMapper extends GenericMapper<TagDTO, Tag>{
+//    @Override
+//    void updateProperties(TagDTO dto, @MappingTarget Tag entity);
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/mapper/TagMapper.java:10]: refactored

File: src/main/java/com/petstore/service/impl/PetServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Transactional(readOnly = true)
+    @Override
+    public PetDTO findById(Long id) {
+        return petMapper.toDTO(petRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(String.format(PET_NOT_FOUND, id))));
+    }
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/service/impl/PetServiceImpl.java:56]: done

File: src/main/java/com/petstore/entity/Category.java [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Entity
+@Table(name = "categories")
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/entity/Category.java:17]: added

File: src/main/java/com/petstore/entity/Orders.java [hunk 15]
```diff
@@ -11,35 +11,34 @@
...
 
+    @Column(name = "pet_id", nullable = false)
     private Long petId;
 
+    @Column(name = "quantity", nullable = false)
     private Integer quantity;
 
     @DateTimeFormat
+    @Column(name = "ship_date", nullable = false)
     private String shipDate;
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/entity/Orders.java:29]: Changed

File: src/main/java/com/petstore/entity/Orders.java [hunk 16]
```diff
@@ -11,35 +11,34 @@
...
+@Table(name = "orders")
+public class Orders extends BaseEntity{
     @Id
-    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "orders_seq")
+    @SequenceGenerator(name = "orders_seq", sequenceName = "orders_seq", initialValue = 5)
     private Long id;
 
+    @Column(name = "pet_id", nullable = false)
     private Long petId;
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/entity/Orders.java:22]: changed

File: src/main/java/com/petstore/service/PetService.java [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.web.multipart.MultipartFile;
+
+import java.util.List;
+
+/**
+ * The PetService interface defines methods for managing pets.
+ */
+public interface PetService {
+    /**
+     * Adds a new pet based on the provided PetDTO object.
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/service/PetService.java:13]: added

File: src/main/java/com/petstore/service/PetService.java [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    PetDTO addPet(PetDTO petDTO);
+
+    /**
+     * Updates the pet with the provided information.
+     *
+     * @param petDTO The PetDTO object containing the updated information for the pet.
+     */
+
+    void updatePet(PetDTO petDTO);
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/service/PetService.java:26]: changed

File: src/main/java/com/petstore/service/PetService.java [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+     * @param id The ID of the pet to delete.
+     */
+    void deletePet(Long id);
+
+    /**
+     * Initializes the folder for pet image uploads.
+     *
+     * @throws RuntimeException if the folder initialization fails
+     */
+    void init();
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/service/PetService.java:65]: done

File: src/main/java/com/petstore/service/StoreService.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+ * Service interface for managing orders in a store.
+ */
+public interface StoreService {
+    /**
+     * Saves the specified order.
+     *
+     * @param order The order to be saved.
+     * @return The saved order.
+     */
+    Orders saveOrder(Orders order);
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/service/StoreService.java:15]: done

File: src/main/java/com/petstore/service/impl/PetServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+     * {@inheritDoc}
+     */
+    @Transactional
+    @Override
+    public PetDTO addPet(PetDTO petDTO) {
+        var pet = petMapper.toEntity(petDTO);
+        var category = categoryRepository.findById(pet.getCategory().getId())
+                .orElseThrow(() -> new EntityNotFoundException(String.format(CATEGORY_NOT_FOUND, pet.getCategory().getId())));
+        pet.setCategory(category);
+        Set<Tag> tags = tagMapper.toSetEntities(petDTO.getTags());
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/service/impl/PetServiceImpl.java:81]: changed

File: src/main/java/com/petstore/service/impl/PetServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+@Service
+@RequiredArgsConstructor
+public class PetServiceImpl implements PetService {
+
+    private final PetRepository petRepository;
+    private final CategoryRepository categoryRepository;
+    private final TagRepository tagRepository;
+    private final PetMapper petMapper;
+    private final TagMapper tagMapper;
+    private final Path root = Paths.get("uploads");
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/service/impl/PetServiceImpl.java:39]: done

File: src/main/java/com/petstore/dto/UserDTO.java [hunk 23]
```diff
@@ -9,7 +9,7 @@
 @Builder
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/dto/UserDTO.java:9]: BaseDTO doesn't have fields so it isn't necessary.

File: src/main/java/com/petstore/service/impl/PetServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+    /**
+     * {@inheritDoc}
+     */
+    @Transactional(readOnly = true)
+    @Override
+    public List<PetDTO> findPetsByStatus(String status) {
+        try {
+            return petMapper.toListDTOs(petRepository.findAllByStatus(Pet.Status.valueOf(status.toUpperCase())));
+        } catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException(INVALID_VALUE);
```
[the remark this answers is not in the document]
Comment (tonikromik) [reply src/main/java/com/petstore/service/impl/PetServiceImpl.java:67]: I want to send my message. Do I need to make my custom exception?

[the remark this answers is not in the document]
Comment (financeasist) [reply src/main/java/com/petstore/service/impl/PetServiceImpl.java]: in that case custom will be better

[the remark this answers is not in the document]
Comment (financeasist) [reply src/main/java/com/petstore/service/impl/PetServiceImpl.java]: in that case custom will be better

**Review by financeasist — APPROVED** [review]

**Review by financeasist — APPROVED** [review]

**Review by financeasist — APPROVED** [review]

**Pull request merged**